### PR TITLE
Fix SPI DMA write/read for ESP32, ESP32-S2

### PR DESCRIPTION
**Pull request by bjoernQ**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description
On ESP32 and ESP32-S2 something like this didn't work
```rust
        embedded_hal_async::spi::SpiBus::write(&mut spi, &send_buffer)
            .await
            .unwrap();

        embedded_hal_async::spi::SpiBus::read(&mut spi, &mut buffer)
            .await
            .unwrap();
```

As already found in #2059 ESP32 needs some special handling to make it work.
For ESP32-S2 the solution was found by "trial and error" (also the problem there was a bit different: on S2 the first iteration worked but then it stopped working)

#### Testing
I added tests and you can modify `embassy_spi.rs` example to perform `write` and `read` in the loop like above

### CODE REVIEW

**Comment by plaes** [comment]
Yes, this fixes at least my issue on regular ESP32 and I can now send and receive full LoRa P2P packets..

**Comment by bugadani** [comment]
Is there a chance the new test can be merged into one of the existing ones?

**Comment by bjoernQ** [comment]
> Is there a chance the new test can be merged into one of the existing ones?

Unfortunately, we need a different `init` - not sure `embedded-test` supports multiple test modules (which probably would be flashed as a separate executable anyways). There are two which in theory might be useable but then the names of the tests are awkward and misleading .... that's why I went with a separate test. Not too happy with that but the other options don't feel great

File: hil-test/tests/spi_full_duplex_dma_write_read.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    use super::*;
+
+    #[init]
+    fn init() -> Context {
+        let peripherals = esp_hal::init(esp_hal::Config::default());
+
+        let io = Io::new(peripherals.GPIO, peripherals.IO_MUX);
+        let sclk = io.pins.gpio0;
+        let (miso, gpio) = hil_test::common_test_pins!(io);
+        let _gpio = Output::new(gpio.degrade(), Level::High);
```
Comment (bugadani) [inline hil-test/tests/spi_full_duplex_dma_write_read.rs:61]:
```suggestion
        let _gpio = Output::new(gpio, Level::High);
```

File: hil-test/tests/spi_full_duplex_dma_write_read.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
+//! SPI Full Duplex DMA write + read Test
+//! See issue #2059
+//!
+//! Folowing pins are used:
+//! SCLK    GPIO0
+//! MISO    GPIO2 / GPIO9  (esp32s2 / esp32s3) / GPIO26 (esp32)
```
Comment (bugadani) [inline hil-test/tests/spi_full_duplex_dma_write_read.rs:6]: I asked Juraj not to introduce this weird separate grouping for esp32, and I'll ask you, too.

Comment (bugadani) [reply hil-test/tests/spi_full_duplex_dma_write_read.rs:6]: Yes. I would like to avoid multiplying bad patterns by copy-paste, please.

Comment (bjoernQ) [reply hil-test/tests/spi_full_duplex_dma_write_read.rs:6]: but seriously does listing the pins in each test make any sense anymore? I don't think so (latest when someone needs to change the pins in lib.rs)

Comment (bugadani) [reply hil-test/tests/spi_full_duplex_dma_write_read.rs:6]: @JurajSadel thinks they do, I don't really agree. And updating them will be a nightmare.

**Review by bugadani — APPROVED** [review]

File: hil-test/tests/spi_full_duplex_dma_write_read.rs (hunk 2 above)
Comment (JurajSadel) [reply hil-test/tests/spi_full_duplex_dma_write_read.rs]: I will remove them from test's description.

**Review by JurajSadel — APPROVED** [review]
Thanks!

**Pull request merged**